tests/test_hello_world.py

```diff
@@ -69,41 +70,140 @@ def test_fibonacci_using_fixture(first_fibonacci_numbers):
         assert fibonacci(item_index) == fibonacci_value
 
 
-# TODO test this function, make sure for example please_test_me("testing is great") = "testing is great!!!"
 def please_test_me(string: str) -> str:
     return string + "!!!"
 
 
+@pytest.mark.parametrize("input_sentence, output_sentence", [("testing is great", "testing is great!!!")])
+def test_please_test_me(input_sentence, output_sentence):
+    assert please_test_me(input_sentence) == output_sentence
+
+
 def times_7(number: Union[int, float]):
     return number * 7
 
 
-# TODO make_me_2_functions_one_use_fixture_and_one_use_parametrize
-def test_make_me_2_functions_one_use_fixture_and_one_use_parametrize():
-    assert times_7(2) == 14
-    assert times_7(4) == 28
-    assert times_7(0) == 0
-    assert times_7(-1) == -7
-    # TODO add one interesting case I didn't check
+@pytest.mark.parametrize("input, output", [(2, 14),(4,28),(0,0),(-1,-7),(0.5,3.5)])
+def test_times_7_parametrize(input, output):
+    assert times_7(input) == output
 
+
+@pytest.fixture
+def first_multiplies_of_7():
+    """
+    This fixture is the first 10 no negative multiplication of 7.
+    """
     random_generator = random.Random()
+    result = []
     for i in range(10):
         rnd_int = random_generator.randint(-1000, 1000)
         # time_7(rnd_int) is like summing 7 items of rnd_int
-        assert times_7(rnd_int) == sum([rnd_int for i in range(7)])
+        result.append((rnd_int,7*rnd_int))
+
+    # Add test for non integer input
+    result.append((0.5,3.5))
+    return result
+
+
+def test_times_7_using_fixture(first_multiplies_of_7):
+    """
+    Test the times_7 function. Tests the first elements of the series.
+    Args:
+        first_multiplies_of_7: This is a fixture so it is automatically filled.
+            The first_multiplies_of_7 will have the first 10 no negative multiplication of 7
+            see first_multiplies_of_7() function.
+    """
+    for test_tupple in first_multiplies_of_7:
+        assert times_7(test_tupple[0]) == test_tupple[1]
+
 
-        # assert times_7(rnd_int) > rnd_int  # TODO Explain why this assert doest work
+def calc_factorial(n: int):
+    if not ((n >= 0) and (int(n) == n)):
```

Comment:
You can is isinatance(n, int) to check n is from type int or n % 1 == 0 to check n is a whole number.

Reply:
Use De-Morgan here ;)
I find this much clearer
```suggestion
    if n < 0 or n % 1 != 0:
```